### ISSUE TITLE
Relationship to HTML

### ISSUE THREAD

**Issue by marcoscaceres** [w3c/contentEditable#2]
Hi @johanneswilm, as I was preparing the document for FPWD I noticed that HTML defines the HTML "contenteditable" attribute itself as well as the "true", "false", "". 

HTML also defines `isContentEditable` IDL attribute. 

It seems we have some duplication between the two specs, which we should try to avoid.

From my own testing, it seems only Chrome supports "plaintext-only". The other specified types don't seem to be supported by other browsers? Or is that incorrect? 

We should probably work out how we should layer these specs or if contentEditable should really just be going directly into the HTML spec. 

@domenic, @annevk, @zcorpan, your guidance as editors of HTML would be greatly appreciated? How would you like these two specs layered (if at all). Should the interoperable parts of the spec just be send to HTML as pull requests?

**Comment by johanneswilm** [comment]
Hey @marcoscaceres yes, so the way we have gone about this is that we first develop the spec and then look at implementations. Because otherwise we end up with more incompatibilities than what we have already. If I have understood you correctly in the past, you think that the specs should be describing how user agents already work than prescribe how they should work. In this case that's not the approach Editing Taskforce took. I believe `plaintext-only` is supported by all browsers that derived from Webkit (so all except Firefox), but it's possible it's not available to the author in Safari currently - I have not checked.

FPWD does not to my knowledge require implementations. Is that correct? There is still a long way to go to get to recommendation status.

I am not opposed to merging into the HTML spec, but that will likely cause a lot of extra work for the maintainers of that spec, as there are almost certain to be changes after FPWD and unless one attends all the Editing Taskforce meetings and follows the email list and github issues, it will be difficult to capture everything that is going there. Maybe merge it at a later date beyond FPWD?

It's true that the `isContentEditable` attribute of elements would likely more logically be in the contenteditable spec, but if it's difficult to change the HTML spec, maybe we should just delete that one sentence about it in the contenteditable spec? Or we leave it in and solve that once we get closer to recommendation?

**Comment by marcoscaceres** [comment]
>  but it's possible it's not available to the author in Safari currently - I have not checked.

Yes, you are correct. It's in Safari and Chrome... though in Chrome it seems to do crazy things. Heh, no wonder we need a spec!  

> FPWD does not to my knowledge require implementations. Is that correct?

That is correct. Yes. Do we have commitments for "event", "caret", and "typing"? 
 
> I am not opposed to merging into the HTML spec, but that will likely cause a lot of extra work for the maintainers of that spec,

"That's why we pay them the big bucks"™️ :) But seriously, the situation we want to avoid is having two sources of truth. Either the HTML is the source of truth or this spec is. 

> as there are almost certain to be changes after FPWD and unless one attends all the Editing Taskforce meetings and follows the email list and github issues, it will be difficult to capture everything that is going there. Maybe merge it at a later date beyond FPWD?

I don't think that needs to be the case. How I imagine it would work is that the Editing Task Force would author the pull requests, then send it to the HTML folks for review. They would help with making sure the layering is done correctly, review the proposed text, and make helpful suggestions... but the work would primarily be done by the task force (HTML editors don't need to follow along). 

> It's true that the isContentEditable attribute of elements would likely more logically be in the contenteditable spec, but if it's difficult to change the HTML spec, maybe we should just delete that one sentence about it in the contenteditable spec? Or we leave it in and solve that once we get closer to recommendation?

I guess it depends on which is the source of truth. 

Let's hear what the HTML folks think and then maybe we can discuss with the larger Editing Task Force? Right now, let's figure out what's best here to avoid duplication.

**Comment by marcoscaceres** [comment]
To jump the shark a little bit... what if this spec only defined "event", "caret", and "typing"? That would avoid the duplication with the HTML spec while allowing the new features to be incubated?

**Comment by johanneswilm** [comment]
We needed to start somewhere. Contenteditable is mentioned a few places in other specs, but there really was no complete approach to it. That's why we started this spec.

I don't think it's a fruitful approach to try to split things out at this point as we are really taking a comprehensive approach to contenteditable and all of these are things we touch on. It would put us into a situation where we suddenly have to get other people in the room because we need to talk about `plaintext-only` for 5 minutes and that type of contenteditable is described in the spec someone unrelated to the editing taskforce is maintaining.

But I do agree that we should make sure things are not in two places when we move to recommendation, and at that stage it may very well make sense just to merge the entire thing into HTML spec instead.

If there is a W3C rule somewhere saying that FPWD specs are not allowed to define something that is already defined in another spec that is a recommendation, then I'd suggest we just add a note "this section is not normative" to all those sections for now and otherwise leave it as it is. If there is no such rule, I would leave it as it is and add a note somewhere in the beginning saying that the HTML spec currently contains some rudimentary definitions related to contentEditable.

**Comment by annevk** [comment]
Usually HTML defines the attribute, but the underlying model is left to other specifications. I think that would work well here too. You can lookup the `referrerpolicy` attribute as an example of this.

**Comment by johanneswilm** [comment]
> Yes, you are correct. It's in Safari and Chrome... though in Chrome it seems to do crazy things. Heh, no wonder we need a spec!

I believe it is used in developer panel. It was an important point for some browser people to get it added to the spec a while back. I forgot which company.

> Do we have commitments for "event", "caret", and "typing"?

These three modes have been developed with the consensus of the participating browser developers - Safari, Chrome and Edge. But we did not obtain a written statement by each browser saying they commit to implement all of them. We actually went back and forth between these modes for several months as we found that these three modes each made sense, but every time we found an issue to tackle, the issue would shift which of these three would make most sense.

I believe it is possible that we eliminate two of the three modes before going to recommendation and even if we keep all three, I would be surprised if all three are implemented simultaneously in all browsers. 

The approach to editing proposed (`EditContext`) by the Edge team at TPAC 2019 would likely be most compatible with "caret". If that proposal falls again, we are likely going to go back to "event" as the main approach because that seems most compatible with the Input Events spec.

> I don't think that needs to be the case. How I imagine it would work is that the Editing Task Force would author the pull requests, then send it to the HTML folks for review. They would help with making sure the layering is done correctly, review the proposed text, and make helpful suggestions... but the work would primarily be done by the task force (HTML editors don't need to follow along).

Given that some important details about this spec will likely change depending on outside factors (such as whether input events or EditContext will be the preferred way to edit text), I think that would result in a lot of extra back and forth. We sort of have something similar between the UI Events spec and the Input Events spec where the UI Events spec took over some aspects, yet text editing isn't their main focus so it's difficult to get their attention some times. I'd prefer we do that when we get closer to recommendation and at that stage we can even merge the entire thing in, if the HTML spec authors prefer to have it organized that way.

> Usually HTML defines the attribute, but the underlying model is left to other specifications. I think that would work well here too. You can lookup the referrerpolicy attribute as an example of this.

@annevk If HTML can define the existence of the attribute and then this spec defines the possible attribute values and the rest of it, I think that should work and not cause too much back and forth.

**Comment by annevk** [comment]
Values are typically defined in HTML too. Still shouldn't cause too much churn I think.

**Comment by johanneswilm** [comment]
ok, in that case if this has to be resolved before FPWD, I wonder if it would be better to wait with FPWD for another 12-48 months and see if we have then have something that can go pretty much straight to rec so we don't have to change the HTML spec more than once for this.

**Comment by annevk** [comment]
My recommendation would be to clearly outline in your document what changes you plan to make to HTML and file an issue against whatwg/html to communicate that. Then once there's tests and implementations are starting we can formalize the remainder. That will help ensure nobody ends up being surprised and there's less chance of others interfering with your intended layering.

**Comment by johanneswilm** [comment]
Ok, so we would file an issue but not resolve it for a while, so that we can make changes to our change request before it actually gets into the HTML spec? A change would be that we drop one or two of the new three values for the contenteditable attribute or that we add a fourth one.

If that is what you are proposing, @annevk , then I would agree that we should go with that.

**Comment by annevk** [comment]
Yeah, the issue would also help with notifying the HTML community, so people who are interested know where to follow-up on things. Thanks!

**Comment by domenic** [comment]
In general I think that a tracking issue as @annevk mentions is important, as well as following the model where HTML defines any content attributes and their values (while potentially delegating the processing model to other specifications).

Given the intention in https://github.com/w3c/contentEditable/issues/2#issuecomment-546825998 to use this as a sort of incubation repository (i.e. speccing features without implementations), versus HTML's approach of working more in tandem with implementations, it probably makes sense to slowly send pull requests to HTML as features or attribute values get more stable and ready to be implemented.

**Comment by marcoscaceres** [comment]
This sounds great. Ok, let's do as suggested above.

**Comment by saschanaz** [comment]
Is the consensus that we keep this spec as an incubator and want to merge fully incubated features into HTML?

AFAICT there is no fully incubated features to merge now, HTML currently just describes the basic of `contentEditable=true/false/inherit` and yet we have [no full description of their behavior](http://w3c.github.io/editing/contentEditableTrue.html).

**Comment by johanneswilm** [comment]
@saschanaz I would personally think it makes sense to keep the contenteditable spec separate as otherwise basically everything on the web could be merged into the HTML spec, but ultimately I would let the maintainers of the HTML spec decide if they want to take on long-term maintainership and merge it into their spec once it's fully developed. I would be OK with it if they actually have the time and resources needed to maintain it.

The spec draft you are linking to is something entirely different, btw. The one you are linking to has a big disclaimer on top saying this will not ever likely move beyond draft stage.

**Comment by marcoscaceres** [comment]
> Is the consensus that we keep this spec as an incubator and want to merge fully incubated features into HTML?

Yes. And we should remove any duplication. 

> and yet we have no full description of their behavior.

Absolutely. That’s what we need to actually specify :)

**Comment by saschanaz** [comment]
>I would personally think it makes sense to keep the contenteditable spec separate as otherwise basically everything on the web could be merged into the HTML spec

Agreed. BTW, what's the standard when maintainers decide to merge a spec into HTML? 👀

>The spec draft you are linking to is something entirely different, btw. The one you are linking to has a big disclaimer on top saying this will not ever likely move beyond draft stage.

Right, but maybe we still need to get some proper descriptions about `contentEditable=true`?

**Comment by johanneswilm** [comment]
> > Is the consensus that we keep this spec as an incubator and want to merge fully incubated features into HTML?
> 
> Yes. And we should remove any duplication.

That's not exactly how I understood it.



> > and yet we have no full description of their behavior.
> 
> Absolutely. That’s what we need to actually specify :)

This is new to me. I don't think we will ever fully describe the expected behavior of contenteditable=True. It's a rabbit hole and I don't think we should attempt to write down what dom changes are to happen based on what user input. And also that draft there is not the one we are dealing with here. There is a disclaimer on top of that file and that disclaimer still is valid. That document should just exist as it is.

**Comment by johanneswilm** [comment]
> Right, but maybe we still need to get some proper descriptions about `contentEditable=true`?

What do you mean by "proper descriptions"? If you say that we should start to say that ctrl+b adds strong tags, or that if the user hits "enter" the paragraph should be split, etc. I think that is just the start of the rabbit hole I mentioned in my last comment so I don't think we should do that. But I'm not to stop others from trying their luck on it.

**Comment by marcoscaceres** [comment]
I think it would be nice to attempt a few pull requests specifying little bits at a time. Let’s do the low hanging fruit and then figure out the complexity.

**Comment by johanneswilm** [comment]
@marcoscaceres Could you give an example of what could be such a small bit?

**Comment by marcoscaceres** [comment]
E.g., `<span contenteditable=true>`... go!

**Comment by johanneswilm** [comment]
Well the existence of the contenteditable attribute is already in various pre-existing specs [1]. But if you were to start specifying **any** of the associated behavior, I think you'd start going down a rabbit hole. Say you specify what hitting `enter` means the current paragraph splits and put that into the HTML-spec. Then it won't take long until someone to figure out you then also need to specify what happens if the selection is non-collapsed. And someone else notices that in some user agents the caret can be in the middle of an SVG and so what to do then if the user hits "enter". A third person finds out that there is a special case if the caret is inside a table cell with the entire table being inside of a paragraph and a `<strong>`-tag, etc. . It's basically a 1000 year discussion to specify all possible circumstances and exceptions, which is why we decided to not try to specify `contenteditable=true` at all and instead continue with the spec as we currently have it that is pretty complete in what it tries to achieve, but hasn't been fully implemented in browsers yet. I'm sure there will be more discussions on it now going forward giving that input events level 1 has been more or less figured out and some of the new modes could work well together with some of this year's new editing proposals.

[1] https://html.spec.whatwg.org/#contenteditable

**Comment by johanneswilm** [comment]
Ok, I filed an issue with HTML and added a temporary note [1] in the abstract of the spec saying what changes it will lead to for the HTML spec. Is this what you had in mind @annevk ?

[1] https://github.com/w3c/contentEditable/pull/4

**Comment by annevk** [comment]
Thanks, that helps.

**Comment by saschanaz** [comment]
I guess ae454fc84c784cd507a1e2f9a226d01f7ac519d4 fixed the duplication: the spec is now empty. 

Hmm, maybe there should be a draft PR to add those removed things again? @marcoscaceres Thoughts?

**Comment by marcoscaceres** [comment]
yes, sorry, the idea was to incrementally add things back in but avoid the duplication. The model that we discussed above still seems ideal. @saschanaz do you have cycles to help us with this?

**Comment by saschanaz** [comment]
Well, given that the spec only defined things that wasn't available on Firefox, not sure what should be done here.

**Comment by johanneswilm** [comment]
The background for this is that Benjamin Peters of Microsoft, who was the central person in the editing taskforce back then, started this spec as part of an initiative to create a new and still to be fully specified lighter alternative to the existing `contenteditable=true`. The entire document only consisted of an idea of what a basic mode could look like, but it was not something that any browser maker had implemented anywhere and was only to serve as a document documenting the current state of discussion on what that would be. Then after some time `contenteditable=plaintext-only` was added (and later removed and even later added again), which made it a bit unclearer  what the document was, because that mode was implemented already.

I am listed as the editor of the spec, but I had no part in any of this initial work. I merely took over most of the documents that Benjamin Peters abandoned when he left. I did continue to update the document as the discussion evolved. It has been some years since we last had a discussion on any of the modes listed in that document.

**Comment by saschanaz** [comment]
Thanks for the background details. The lack of recent discussion sounds like there is no active implementor interest, though. Maybe just wait and see where EditContext goes? 🤔

**Comment by johanneswilm** [comment]
Yes, so we put in on the shelve for after Input Events. And now it seems like EditContext in a sense can work as a replacement for a lot of what a lot of people hoped Input Events level 2 would do, so waiting until then is probably a good idea. I do think though it would be good to bring back the text in some form (a PR or otherwise), as I can see there are people out there looking for it and they start using the wayback machine and other tools to get to it. See for example the discussion starting with this comment: https://github.com/w3c/editing/issues/162#issuecomment-615941294
